include/faabric/mpi/mpi.h

```diff
@@ -70,7 +70,11 @@ extern "C"
         int worldId;
         int rank;
         int size;
-        int wasmPtr;
+        union
```

Comment:
This is the only change I've needed to make to the original interface (kudos to you for that). The reason being that we are now dealing with real memory and not offsets in the WASM memory array.

Reply:
A `union` is nice because it means there's only one set of code shared between both compilation targets, but I'm not sure it's the most common way to change things based on compilation target. I would say it's more normal to use preprocessor directives with things like `#ifdef __wasm__`. In this case you can kind of guess that one gets used in wasm and the other doesn't based on the names I guess.

I would have expected a type called something like `MPI_POINTER_TYPE` with an `#ifdef __wasm__` around it to decide whether it's `int` or `void*`. It's more explicit that way but if you add a comment to explain the union then that's ok too.

Reply:
I agree that preprocessor directives are a better choice for things based on compilation targets. I don't think though that the `MPI_POINTER_TYPE` is needed as this only happens once.

I've used CMake's `add_compile_definitions` to add a preprocessor directive if we detect a `Wasm` MPI build.

Reply:
I don't think we need to use an extra compile definition do we? Can we just use `#ifdef __wasm__`?

Reply:
I wasn't aware `wasi` set this value somewhere upstream (I literally redefined `__wasm__`).